src/test/peerfinder/Livecache_test.cpp

```diff
@@ -22,49 +22,190 @@
+        BEAST_EXPECT(c.size() == 10);
+
+        for (auto i = 0; i < 10; ++i)
+            add(beast::IP::randomEP(false), c);
+
+        BEAST_EXPECT(! c.empty());
+        BEAST_EXPECT(c.size() == 20);
     }
 
-    void testFetch ()
+    void testInsertUpdate ()
     {
+        testcase ("Insert/Update");
         Livecache <> c (m_clock, beast::Journal());
 
-        add (1, 1, c);
-        add (2, 1, c);
-        add (3, 1, c);
-        add (4, 1, c);
-        add (4, 2, c);
-        add (4, 3, c);
-        add (5, 1, c);
-        add (6, 1, c);
-        add (6, 2, c);
-        add (7, 1, c);
+        auto ep1 = Endpoint {beast::IP::randomEP(), 2};
+        c.insert(ep1);
+        BEAST_EXPECT(c.size() == 1);
+        // third position list will contain the entry
+        BEAST_EXPECT((c.hops.begin()+2)->begin()->hops == 2);
+
+        auto ep2 = Endpoint {ep1.address, 4};
+        // this will not change the entry has higher hops
+        c.insert(ep2);
+        BEAST_EXPECT(c.size() == 1);
+        // still in third position list
+        BEAST_EXPECT((c.hops.begin()+2)->begin()->hops == 2);
+
+        auto ep3 = Endpoint {ep1.address, 2};
+        // this will not change the entry has the same hops as existing
+        c.insert(ep3);
+        BEAST_EXPECT(c.size() == 1);
+        // still in third position list
+        BEAST_EXPECT((c.hops.begin()+2)->begin()->hops == 2);
+
+        auto ep4 = Endpoint {ep1.address, 1};
+        c.insert(ep4);
+        BEAST_EXPECT(c.size() == 1);
+        // now at second position list
+        BEAST_EXPECT((c.hops.begin()+1)->begin()->hops == 1);
+    }
+
+    void testExpire ()
+    {
+        testcase ("Expire");
+        using namespace std::chrono_literals;
+        Livecache <> c (m_clock, beast::Journal());
+
+        auto ep1 = Endpoint {beast::IP::randomEP(), 1};
+        c.insert(ep1);
+        BEAST_EXPECT(c.size() == 1);
+        c.expire();
+        BEAST_EXPECT(c.size() == 1);
+        // verify that advancing to 1 sec before expiration
+        // leaves our entry intact
+        m_clock.advance(Tuning::liveCacheSecondsToLive - 1s);
+        c.expire();
+        BEAST_EXPECT(c.size() == 1);
+        // now advance to the point of expiration
+        m_clock.advance(1s);
+        c.expire();
+        BEAST_EXPECT(c.empty());
+    }
+
+    void testHistogram ()
+    {
+        testcase ("Histogram");
+        Livecache <> c (m_clock, beast::Journal());
+        for (auto i = 0; i < 40; ++i)
+            add(
+                beast::IP::randomEP(true),
+                c,
+                ripple::rand_int(0, static_cast<int>(Tuning::maxHops + 1)));
+        auto h = c.hops.histogram();
+        if(! BEAST_EXPECT(! h.empty()))
+            return;
+        std::vector <std::string> v;
+        boost::split (v, h, boost::algorithm::is_any_of (","));
+        for (auto const& n : v)
+            BEAST_EXPECT(boost::lexical_cast<int>(boost::trim_copy(n)) >= 0);
```

Comment:
Since you're converting to `int` anyway, why not sum them up and check that they add up to 40.